### PR TITLE
Browser painting initial implementation

### PR DESCRIPTION
**Pull request by PlasmaFlare**
Initial draft for #6 

Features:
- Adds toolbar on left for terrain editing
- Support for brush, fill, and eyedropper tools
- Palette selection
- Change brush size

Notes:
- Brush rendering uses a buffer array/overlay to render the drawn pixels in a stroke. When mouse is released, the buffer contents are committed to main geo, and buffer gets flushed. This is done to reduce overhead of constantly decoding and encoding geo per pixel.

### CODE REVIEW

File: src/levelEditor/toolbar/LevelToolbar.jsx [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+}>;
+
+export default function LevelToolbar(props: Props): React$Node {
+	const toolTypes = ['Select', 'Brush', 'Fill', 'Eyedropper'];
+
+	const [colorDescription, setColorDescription] =
+		useState<?ToolbarColorItem>(null);
+	useEffect(() => {
+		setColorDescription(TOOLBAR_COLOR_LOOKUP.get(props.currentPaintColor));
+	}, [props.currentPaintColor]);
```
Comment (ngyikp) [inline src/levelEditor/toolbar/LevelToolbar.jsx:32]: No need for state
```suggestion
	const colorDescription = TOOLBAR_COLOR_LOOKUP.get(props.currentPaintColor);
```

File: src/levelEditor/toolbar/LevelToolbar.jsx [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+					}}
+				/>
+			</div>
+			Palette Selection:
+			{PIXEL_COLORS_EXPLANATIONS.map((color, _) => {
+				return color.colors.map((colorIndex, arrayIndex) => {
+					let description = color.description;
+					if (description === 'Higher ground layers') {
+						description = 'Height ' + arrayIndex;
+					}
```
Comment (ngyikp) [inline src/levelEditor/toolbar/LevelToolbar.jsx:78]: Isn't this from TOOLBAR_COLOR_LOOKUP?

File: src/levelEditor/util/drawGeoToCanvas.js [hunk 3]
```diff
@@ -20,7 +22,12 @@ export default function drawGeoToCanvas({
...
 	geo.forEach((pixel, index) => {
-		const fill = PIXEL_COLORS.get(pixel);
+		let pixelToUse = pixel;
+		if (geoPaintBuffer != null && geoPaintBuffer[index] != null) {
+			pixelToUse = geoPaintBuffer[index];
+		}
+
+		const fill = PIXEL_COLORS.get(pixelToUse);
 		if (fill == null) {
 			console.warn('unknown pixel color ' + pixel);
```
Comment (ngyikp) [inline src/levelEditor/util/drawGeoToCanvas.js:32]:
```suggestion
			console.warn('unknown pixel color ' + pixelToUse);
```

**Review by ngyikp — CHANGES_REQUESTED** [review]

File: src/levelEditor/GeoConstants.js [hunk 4]
```diff
@@ -69,3 +69,23 @@ export const PIXEL_COLORS_EXPLANATIONS = [
...
+		let description = colorItem.description;
+		if (description === 'Higher ground layers') {
+			description = 'Height ' + arrayIndex;
+		}
+
+		TOOLBAR_COLOR_LOOKUP.set(color, {
+			description: description,
+		});
+	});
+}
```
[the remark this answers is not in the document]
Comment (PlasmaFlare) [reply src/levelEditor/GeoConstants.js:89]: This is supposed to map color indexes => descriptions based on `PIXEL_COLORS_EXPLANATIONS`. `PIXEL_COLORS_EXPLANATIONS` doesn't really do that by itself.

File: src/levelEditor/LevelInspector.jsx [hunk 5]
```diff
@@ -83,23 +100,151 @@ export default function LevelInspector({
...
+			if (editorToolType === 'Brush' && isPainting) {
+				const rect = ev.currentTarget.getBoundingClientRect();
+
+				const mouseMapCoords = [
+					parseInt(ev.clientX - rect.left, 10),
+					parseInt(ev.clientY - rect.top, 10),
+				];
+
+				paint(mouseMapCoords);
+			}
```
[the remark this answers is not in the document]
Comment (PlasmaFlare) [reply src/levelEditor/LevelInspector.jsx:230]: Without this code, if the user holds down the button while quickly moving the cursor to be outside the geo preview, `onMapMouseMove` will not fire to paint the pixels between `prevCoordinates` and the cursor. We need `onMapMouseLeave` to cover this.

File: src/levelEditor/LevelInspector.jsx [hunk 6]
```diff
@@ -203,6 +348,9 @@ export default function LevelInspector({
 	const onActiveUiViewToggle = useCallback((uiView: LevelInspectorUiView) => {
 		setActiveUiViews((activeUiViews) => {
 			if (activeUiViews.includes(uiView)) {
+				if (uiView === 'GEO') {
+					setEditorToolType('Select');
+				}
```
[the remark this answers is not in the document]
Comment (PlasmaFlare) [reply src/levelEditor/LevelInspector.jsx:359]: ~~THIS IS AN INVALID COMMIT BAN HIIMMM ARRRRRRGGGGG~~

(nah jk, discussed with ng and agreed to give more thought)

File: src/levelEditor/common/GeoPreview.jsx [hunk 7]
```diff
@@ -20,6 +20,8 @@ type Props = $ReadOnly<{
 	mapMouseMoveCoordinates: ?[number, number],
 	scale: number,
 	useDevicePixelRatio: boolean,
+	geoPaintBuffer: ?Array<number>,
+	paintBufferUpdate?: ?number,
```
[the remark this answers is not in the document]
Comment (ngyikp) [reply src/levelEditor/common/GeoPreview.jsx:24]: Turns out `geoPaintBuffer` is a mutable array so the array reference would remain the same

**Comment by ngyikp** [comment]
pizza dance

**Pull request merged**